### PR TITLE
DAOS-5964 control: autoconfig generate target parameters

### PR DESCRIPTION
**Pull request by tanabarr**
Detect recommended values for target count and helper xstream thread
count parameters during automatic configuration file generation.
Refactor logic in control API and algorithm for selecting recommended
values to improve clarity and coverage of different machine architectures.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/lib/control/auto.go [hunk 1]
```diff
@@ -228,29 +234,33 @@ func (req *ConfigGenerateReq) genConfig(interfaces []*HostFabricInterface) *conf
 
 // checkNetwork scans fabric network interfaces and updates-in-place configuration
 // with appropriate fabric device details for all required numa nodes.
-func (req *ConfigGenerateReq) checkNetwork(ctx context.Context, resp *ConfigGenerateResp) error {
+//
+// Updates response config field in-place and returns number of cores per NUMA
+// node, which is the same on all hosts in the host set.
+func (req *ConfigGenerateReq) checkNetwork(ctx context.Context, resp *ConfigGenerateResp) (int, error) {
```
Comment (mjmac) [inline src/control/lib/control/auto.go:240]: It's kind of weird that a function named `checkNetwork()` is returning the number of cores per NUMA node, isn't it?

Out of curiosity, why not use the `netdetect.CoresPerNumaNode()` helper that you added?

**Review by jbrosenz — APPROVED** [review]
At first, I didn't see why adding the CoresPerNumaNode made sense architecturally, but then I saw how you were using hwloc services to give you the data you needed and it made more sense.  I could imagine that argument that we should refactor netdetect so that hwloc is no longer part of it, but I don't advocate that at this time.  Looks good for the intended purpose and looks like a reasonable solution.

File: src/control/lib/control/auto.go [hunk 2]
```diff
@@ -449,3 +475,55 @@ func (req *ConfigGenerateReq) checkStorage(ctx context.Context, resp *ConfigGene
...
+		switch ioCfg.Storage.Bdev.DeviceCount {
+		case 1, 2, 4, 8:
+			numTargets = 16
+		case 5:
+			numTargets = 15
+		case 7:
+			numTargets = 14
+		case 3, 6:
+			numTargets = 12
+		}
```
Comment (mjmac) [inline src/control/lib/control/auto.go:505]: Seems like there should be a default case here. What happens if there are more devices than 8?

**Review by mjmac — COMMENTED** [review]
Approach looks good, just a few comments.

File: src/control/lib/control/auto.go (hunk 1 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:240]: That would give you the cores on the client (the node that dmg is being run on), the helper needs to be run on the server which happens in NetworkScan request handler. Make sense?

File: src/control/lib/control/auto.go (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:505]: I removed the specific validation from this function because it already happens in validateNvmeStorage() https://github.com/daos-stack/daos/pull/3889/files/1677c4c990ef584e579e38230c23fb24de31ecf6#diff-a723789e4c16a154f22560c66bff2424c7b75c78231333a593aa878cbcf330d8R377

Comment (kjacque) [reply src/control/lib/control/auto.go:505]: If it can't happen, maybe the default should return an error. I was going to suggest a panic, but that might get us in more trouble if we make a coding error in the future and tests don't catch it.

File: src/control/lib/control/auto.go (hunk 1 above)
Comment (kjacque) [reply src/control/lib/control/auto.go:240]: I understand why this is where the number of cores per NUMA node is coming from, but it still feels surprising to return it here. I think a minor refactoring would make things clearer. Here is one rough suggestion, you may be able to think of something better:

- add a new method req.buildConfig(resp *GenerateConfigResp) to encapsulate all of the logic:
  - req.getSingleNetworkSet
  - req.checkNetwork (now accepts req and fabricInfo, doesn't call getSingleNetworkSet itself anymore)
  - req.checkStorage
  - req.checkCPUs (fabricInfo.CoresPerNumaNode can now be passed in directly)

**Review by kjacque — COMMENTED** [review]
Some suggestions relating to @mjmac's comments, otherwise looks reasonable to me.

File: src/control/lib/control/auto.go (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:505]: Would be good to have some clarification on this as we have some inconsistencies in the code base regarding the use of panic(), I noticed one was just introduced in the multi-replica PR. Sometimes when a situation "really shouldn't happen" we return an error and occasionally we panic(). Obviously we don't want to panic too much but should we be panicking at all?

File: src/control/lib/control/auto.go (hunk 1 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:240]: this may make things clearer but the reason I didn't do this is that it makes testing awkward, mock invoker currently doesn't support returning different types of responses as far as I know so in the above example we will have to add another test for getSingleNetworkSet, but I can go down that route if preferred

**Review by kjacque — APPROVED** [review]
LGTM

File: src/control/lib/control/auto_test.go [hunk 3]
```diff
@@ -519,25 +549,126 @@ func TestControl_AutoConfig_checkNetwork(t *testing.T) {
...
+		"18 cores 16 ssds":   {18, []int{16}, []int{16}, []int{1}, nil},
+		"16 cores no ssds":   {16, []int{0}, []int{16}, []int{1}, errCores(0, 17, 16)},
+		"16 cores 1 ssds":    {16, []int{1}, []int{16}, []int{1}, errCores(1, 17, 16)},
+		"16 cores 2 ssds":    {16, []int{2}, []int{16}, []int{1}, errCores(2, 17, 16)},
+		"16 cores 3 ssds":    {16, []int{3}, []int{12}, []int{3}, nil},
+		"16 cores 4 ssds":    {16, []int{4}, []int{16}, []int{1}, errCores(4, 17, 16)},
+		"16 cores 5 ssds":    {16, []int{5}, []int{15}, []int{0}, nil},
+		"16 cores 6 ssds":    {16, []int{6}, []int{12}, []int{3}, nil},
+		"16 cores 7 ssds":    {16, []int{7}, []int{14}, []int{1}, nil},
+		"16 cores 8 ssds":    {16, []int{8}, []int{16}, []int{1}, errCores(8, 17, 16)},
```
Comment (kjacque) [inline src/control/lib/control/auto_test.go:624]: These failure cases seem so odd to me, it makes it feel like folks with a 16 core CPU are out of luck. Couldn't we drop back to the next smallest multiple of the number of SSDs, even if it's less performant?

**Review by kjacque — DISMISSED** [review]
Nothing I'd block on, but I have a couple of comments to think about for follow-on improvements. I wonder if the numTargets calculations could be simplified to a formula instead of depending on a switch statement for different numbers of SSDs.

File: src/control/lib/control/auto_test.go (hunk 3 above)
Comment (tanabarr) [reply src/control/lib/control/auto_test.go:624]: I have been following guidance from @johannlombardi as documented in the ticket, it's expected that most installations will have sockets with 18 physical cores or more so the 16 and 32 core tests are edge cases, common use cases will be 18 & 24. Happy to adjust the selection if wanted.

File: src/control/lib/control/auto.go (hunk 2 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:518]: yes I agree with that and it's the approach I've been following

File: src/control/lib/control/auto_test.go [hunk 4]
```diff
@@ -519,25 +549,126 @@ func TestControl_AutoConfig_checkNetwork(t *testing.T) {
...
+
+	for name, tc := range map[string]struct {
+		numCores      int
+		bdevCounts    []int // one for each I/O Server in resp.ConfigOut
+		expTgtCounts  []int
+		expHlprCounts []int
+		expCheckErr   error
+	}{
+		"no cores":           {expCheckErr: errors.Errorf(errInvalNumCores, 0)},
+		"24 cores no ssds":   {24, []int{0}, []int{16}, []int{7}, nil},
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/auto_test.go:596]: that's what I had originally but it takes up a huge amount of lines so I changed to this format, it's hard to see the coverage on a single page if it's in the verbose format

**Comment by tanabarr** [comment]
> Nothing I'd block on, but I have a couple of comments to think about for follow-on improvements. I wonder if the numTargets calculations could be simplified to a formula instead of depending on a switch statement for different numbers of SSDs.

I wanted to represent in a formula , but given the number of special cases required (see the ticket description and comments for more details) that wasn't practical. Happy to work on fleshing out a more general formula if that's agreeable.

File: src/control/lib/control/auto_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/control/auto_test.go:596]: That's fair. There are a lot of cases here.

File: src/control/lib/control/auto_test.go (hunk 3 above)
Comment (kjacque) [reply src/control/lib/control/auto_test.go:624]: I'd assume Johann knows more about the use cases than I do. If it's possible to have 16 cores, though, my instinct would be to produce some kind of working config for them, even though it's not as performant. May not be a high priority though.

**Comment by kjacque** [comment]
> I wanted to represent in a formula , but given the number of special cases required (see the ticket description and comments for more details) that wasn't practical. Happy to work on fleshing out a more general formula if that's agreeable.

I don't think it's urgent, so maybe we could deal with it in some future PR. It seems like the general algorithm is multiply the number of SSDs until you get as close as possible to 16 (over or under, but preferring under if it's between 12-16), and then drop back by one multiple if you don't have enough cores.

**Review by mjmac — COMMENTED** [review]
I know that it's late to be giving this particular feedback, but I have to say that I'm not really happy with the current implementation. When I approved the first PR in this series, I did it without checking out the code to examine it directly and really get a good understanding of how it was laid out, and I'm now regretting that. Putting all of the logic in methods on the request struct makes this a lot harder to understand and test, IMO. There are tons of non-apparent side effects on the supplied response struct and things returned in surprising ways.

I'm not going to insist on it for this PR, but in the near future I think it would be a good idea to revisit this code to refactor for clarity and testability. Instead of putting all of the logic in methods on the request struct, it would be much better to rework them as standalone functions which accept relevant parameters from the request (or the request itself, I suppose, as long as there are no modifications made) and return items which are added to the response.

Among other issues, it's confusing to have methods named `check*()` that make modifications to the response (as an example of what I mean by side effects).

**Comment by tanabarr** [comment]
> > I wanted to represent in a formula , but given the number of special cases required (see the ticket description and comments for more details) that wasn't practical. Happy to work on fleshing out a more general formula if that's agreeable.
> 
> I don't think it's urgent, so maybe we could deal with it in some future PR. It seems like the general algorithm is increase multiples of the number of SSDs until you get as close as possible to 16 (over or under, but preferring under if it's between 12-16), and then drop back by one multiple if you don't have enough cores.

The algorithm you suggest doesn't work for the criteria specified for 9 or 10 SSDs (as detailed in the ticket), if requirements were changed slightly , then we could apply a simple formula.

File: src/control/lib/control/auto.go [hunk 5]
```diff
@@ -449,3 +468,76 @@ func (req *ConfigGenerateReq) checkStorage(ctx context.Context, resp *ConfigGene
...
+//
+// Create helper threads for the remaining available cores, e.g. with 24 cores,
+// allocate 7 helper threads. Number of helper threads should never be more than
+// number of targets.
+func (req *ConfigGenerateReq) checkCPUs(coresPerNuma int, resp *ConfigGenerateResp) error {
+	if coresPerNuma < 1 {
+		return errors.Errorf(errInvalNumCores, coresPerNuma)
+	}
+
+	for idx, ioCfg := range resp.ConfigOut.Servers {
```
[the remark this answers is not in the document]
Comment (tanabarr) [reply src/control/lib/control/auto.go:490]: I will generalise the algorithm, the current implementation is not general because specific special cases were requested that didn't fit into a general algorithm, the ranges selected matched what was requested.

**Comment by tanabarr** [comment]
> I know that it's late to be giving this particular feedback, but I have to say that I'm not really happy with the current implementation. When I approved the first PR in this series, I did it without checking out the code to examine it directly and really get a good understanding of how it was laid out, and I'm now regretting that. Putting all of the logic in methods on the request struct makes this a lot harder to understand and test, IMO. There are tons of non-apparent side effects on the supplied response struct and things returned in surprising ways.
> 
> I'm not going to insist on it for this PR, but in the near future I think it would be a good idea to revisit this code to refactor for clarity and testability. Instead of putting all of the logic in methods on the request struct, it would be much better to rework them as standalone functions which accept relevant parameters from the request (or the request itself, I suppose, as long as there are no modifications made) and return items which are added to the response.
> 
> Among other issues, it's confusing to have methods named `check*()` that make modifications to the response (as an example of what I mean by side effects).

The timing of this review feedback is disappointing given that you have had significant time to review the code, I implemented this way to reduce the number of variables passed as parameters because I was trying to pre-empt feedback from you which I anticipated (incorrectly) would advise me against passing numerous variables to and from each function. I will revise.

**Comment by kjacque** [comment]
> The algorithm you suggest doesn't work for the criteria specified for 9 or 10 SSDs (as detailed in the ticket), if requirements were changed slightly , then we could apply a simple formula.

With 9 or 10 SSDs it still works, I think. The double values (18 or 20) should be picked if there are enough cores to support them, since both are closer to 16 than 9 or 10 are.

Looking over the cases, 3 actually doesn't work, since the value suggested is 12 instead of 15. But maybe 15 would work just as well? It might be something to chat with @johannlombardi about.

File: src/control/lib/control/auto.go [hunk 6]
```diff
@@ -477,6 +477,22 @@ func checkStorage(log logging.Logger, numPmem, reqNumNvme int, storageSet *HostS
 	return pmemPaths, bdevLists, nil
 }
 
+func calcHelpers(log logging.Logger, targets, cores int) int {
+	helpers := cores - targets - 1
```
Comment (jbrosenz) [inline src/control/lib/control/auto.go:481]: Is there a need to do any validation to the targets and cores to make sure they are positive?

**Review by jbrosenz — APPROVED** [review]
Looks good to me.  I had one question, but it shouldn't block.

File: src/control/lib/control/auto.go (hunk 6 above)
Comment (tanabarr) [reply src/control/lib/control/auto.go:481]: Whilst it's probably sensible to future proof the code by adding the check, in both the calls to this function it would be impossible for helpers to end up negative (cores > targets guaranteed in both calls).

**Comment by tanabarr** [comment]
CI failures:
- DAOS-5807 OSAOnlineReintegration
- DAOS-6121 MdtestSmall
- DAOS-5591 MacsioTest
- DAOS-6025 DAOS vc tests
- DAOS-6111 functional Leap15 suite

File: src/control/lib/netdetect/netdetect.go [hunk 7]
```diff
@@ -288,6 +289,15 @@ func Init(parent context.Context) (context.Context, error) {
 	}
 	ndc.numNUMANodes = numNUMANodes(ndc.topology)
 	ndc.numaAware = ndc.numNUMANodes > 0
+	if ndc.numaAware {
+		cores, err := getCoreCount(ndc.topology)
+		if err != nil {
+			return nil, err
+		}
+		ndc.coresPerNuma = cores / ndc.numNUMANodes
```
Comment (mjmac) [inline src/control/lib/netdetect/netdetect.go:297]: I think this is a fine simplifying assumption, but I wonder if it will always be true. I could imagine some weird hardware with unbalanced sockets (mixture of on-die CPUs and GPUs or something like that). Probably not something we're likely to encounter on storage nodes, though.

**Review by mjmac — APPROVED** [review]

File: src/control/lib/netdetect/netdetect.go (hunk 7 above)
Comment (tanabarr) [reply src/control/lib/netdetect/netdetect.go:297]: Agreed, this assumption won't hold for architectures with asymmetric CPU design across sockets.
